[14주차] 심규창 학습 PR 제출합니다.

gyuchangShim opened the pull request:
Postman: https://documenter.getpostman.com/view/30531137/2s9YsDjaB2
ERD: ![jpa_practice](https://github.com/COW-edu/jpa-practice/assets/132640569/b544c15a-2073-403f-b3af-6e2a125f97c9)

gyuchangShim committed:
Feat: Member CRUD 기능 추가

gyuchangShim committed:
Feat: Board CRUD 기능 추가

gyuchangShim committed:
Feat: Reply CRUD 기능 추가

gyuchangShim committed:
Fix: Board & Reply 조회 시 내용 변경

stophwan commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+        this.title = boardUpdateRequest.getTitle();
+        this.content = boardUpdateRequest.getContent();
+        this.date = LocalDate.now();
+    }
```
도에인에 Request 객체가 있는 형태는 좋지 않은 방식입니다!
가장 큰 이유는 둘이 사이클 차이가 너무 큽니다. 변경 가능성이 많이 없는 Board 객체에 비해서 웹에 가장 앞에 있는 Request 객체는 변경 가능성이 높거든요. 요청이 변경될 때마다 도메인 객체를 변경해야하니까요.

stophwan commented on src/main/java/com/example/jpa/board/dto/request/BoardCreateRequest.java:
```diff
@@ -0,0 +1,15 @@
+
+    private String title;
+    private String content;
+}
```
Reqeust에서 검증 조건이 너무 없습니다. 당장 id나 title content가 null 값이 들어와도 그대로 다 받아버려요.
Column은 JPA에서 사용하는 어노테이션으로 알고 있는데 불필요해 보입니다.

gyuchangShim replied:
Spring validation dependency 사용해서 검증 조건 추가했습니다!

stophwan commented on src/main/java/com/example/jpa/member/service/MemberService.java:
```diff
@@ -18,16 +23,46 @@ public class MemberService {
+                throw new NullPointerException("이미 가입한 사용자입니다.");
+            }
+
+        }
```
member가 null인지 아닌지 여기서 체크하는 것보다는 애당초 Repository에서 Optional로 받아 처리하는 것을 어떨까요?
member.getPhoneNumber() 보다는 Member 도메인 내에서 처리하는 것이 더 좋아보여요.
이미 가입한 사용자라는 에러를 반환하는데 있어 NullPointException은 올바르지 않아보입니다!

gyuchangShim replied:
null 객체에 대한 확인은 피드백 반영해서 Optional로 처리했습니다! 그리고 검증 과정은 Member 도메인에서 진행하고 CustomException을 만들어 에러 반환하도록 변경했습니다.

stophwan commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+@Getter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
지양해야하는 패턴입니다. 이러면 private필드로 설정한 이유가 없습니다. 가벼운 토이프로젝트에서는 상관 없으나 실제 프로젝트에서는 주의해주세요!

dojinyou replied:
`@Builder`와 `@Setter`가 정말 필요한지 고민해보기

k-kbk commented on src/main/java/com/example/jpa/board/controller/BoardController.java:
```diff
@@ -0,0 +1,49 @@
+        return ResponseEntity.ok().build();
+    }
+
+    @PutMapping
+    public ResponseEntity<Void> updateBoard(@RequestBody BoardUpdateRequest boardUpdateRequest) {
+        boardService.updateBoard(boardUpdateRequest);
+        return ResponseEntity.ok().build();
+    }
```
URL에 식별자를 포함하는게 조금 더 RESTful 할 수도 있을 것 같네요! 하지만 절대적인 것은 아니니 참고만 하시면 될 것 같습니다 👍

k-kbk commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+//        this.date = date;
+//    }
+
+    public static Board of(BoardCreateRequest boardCreateRequest, Member member) {
+        return Board.builder()
+                .member(member)
+                .title(boardCreateRequest.getTitle())
+                .content(boardCreateRequest.getContent())
+                .date(LocalDate.now())
+                .build();
+    }
```
좋은 것 같습니다 👍

k-kbk commented on src/main/java/com/example/jpa/member/domain/Member.java:
```diff
@@ -16,16 +20,29 @@ public class Member {
+    private int age;
+    private String phoneNumber;
+
+    @OneToMany(mappedBy = "member", orphanRemoval = true)
```
👍

k-kbk commented on src/main/java/com/example/jpa/member/dto/request/MemberCreateRequest.java:
```diff
@@ -7,10 +7,14 @@
+    private int age;
+    private String phoneNumber;
 
     public Member toEntity() {
         return Member.builder()
                 .name(this.name)
+                .age(this.age)
+                .phoneNumber(this.phoneNumber)
                 .build();
     }
```
👍

k-kbk commented on src/main/resources/application.yml:
```diff
@@ -6,7 +6,7 @@ spring:
     url: jdbc:mysql://localhost:3306/jpa?serverTimezone=UTC
     driver-class-name: com.mysql.cj.jdbc.Driver
     username: root
-    password: mysql
+    password: tlarbckd@17
```
이런 정보는 보안에 매우 중요한 부분이니 환경변수화해서 외부에 감추면 좋을 것 같아요!

k-kbk commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class Board {
```
게시물이다보니 board라는 이름보다는 post같은 이름이 조금 더 자연스러울 것 같네요!

k-kbk commented:
전반적으로 정말 잘 작성하신 것 같아요! 👍
고생하셨어요!

dojinyou commented on src/main/java/com/example/jpa/board/controller/BoardController.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.List;
+
+@RestController
```
`@RestContoller`와 `@Controller` 알아보기!

dojinyou commented on src/main/java/com/example/jpa/board/controller/BoardController.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("api/boards")
```
```suggestion
@RequestMapping("/api/boards")
```

dojinyou commented on src/main/java/com/example/jpa/board/controller/BoardController.java:
```diff
@@ -0,0 +1,49 @@
+        boardService.updateBoard(boardUpdateRequest);
+        return ResponseEntity.ok().build();
+    }
+
```
```suggestion
```

dojinyou commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "member_id")
+    private Member member;
```
작성자의 의미라면 writer or author 같은 이름이 더 적절할 것 같음
```suggestion
    private Member writer;
```

dojinyou commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+    @JoinColumn(name = "member_id")
+    private Member member;
+
+    @OneToMany(mappedBy = "board", orphanRemoval = true)
+    private List<Reply> replyList = new ArrayList<>();
```
`List<Reply>`를 가지고 있는게 좋은 지 고민해보기

dojinyou commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+//        this.title = title;
+//        this.content = content;
+//        this.date = date;
+//    }
```
불필요한 주석 제거
```suggestion
```

dojinyou commented on src/main/java/com/example/jpa/board/domain/Board.java:
```diff
@@ -0,0 +1,59 @@
+
+    private String title;
+    private String content;
+    private LocalDate date;
```
createdDate, UpdatedDate같은 명확한 네이밍 가져가기
```suggestion
    private LocalDate createdDate;
```

dojinyou commented on src/main/java/com/example/jpa/board/dto/request/BoardCreateRequest.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Getter;
+
+@Getter
+public class BoardCreateRequest {
```
`record` 활용해보기

dojinyou commented on src/main/java/com/example/jpa/board/service/BoardService.java:
```diff
@@ -0,0 +1,56 @@
+    private final BoardRepository boardRepository;
+    private final MemberService memberService;
+
+    @Transactional
```
[nit]class 레벨에 적용하고 필요한 부분만 메서드에 따로 작용해도 좋을듯

dojinyou commented on src/main/java/com/example/jpa/board/service/BoardService.java:
```diff
@@ -0,0 +1,56 @@
+        targetBoard.update(boardUpdateRequest);
+    }
+
+    public Board checkExist(Long id) {
+        Board targetBoard = boardRepository.findById(id)
+                .orElseThrow(() -> new NoSuchElementException("게시판이 존재하지 않습니다."));
+        return targetBoard;
+    }
```
chekcExist라면 boolean 응답을 기대할 것 같음. get이나 fetch로 쓰는게 적절할듯

stophwan replied:
도진님 말씀 굉장히 동감합니다. 메서드가 직관적이지 않은것 가타용.

dojinyou commented on src/main/java/com/example/jpa/member/controller/MemberController.java:
```diff
@@ -1,17 +1,15 @@
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
전체 import 지양하기

dojinyou commented on src/main/java/com/example/jpa/member/dto/request/MemberCreateRequest.java:
```diff
@@ -7,10 +7,14 @@
 
     private String name;
+    private int age;
+    private String phoneNumber;
```
```suggestion
    private final String name;
    private final int age;
    private final String phoneNumber;
```

dojinyou commented on src/main/java/com/example/jpa/member/service/MemberService.java:
```diff
@@ -18,16 +23,46 @@ public class MemberService {
+    private void checkDuplicate(MemberCreateRequest request) {
+        Member member = memberRepository.findByName(request.getName());
+        if(member != null) {
+            if(member.getPhoneNumber().equals(request.getPhoneNumber())) {
```
도메인 객체에게 묻기(`hasSamePhoneNumber`)
if 뎁스 줄이기
```suggestion
        if(member != null && member.hasSamePhoneNumber(request.getPhoneNumber()) {
```

gyuchangShim replied:
다른 피드백과 더해서 Optional로 null에 대한 여부 확인과 hasSamePhoneNumber 메서드를 도메인 객체에 생성해 처리했습니다!

dojinyou commented on src/main/java/com/example/jpa/member/service/MemberService.java:
```diff
@@ -18,16 +23,46 @@ public class MemberService {
+
+        }
+
+    }
```
```suggestion
            }
        }
    }
```

dojinyou commented on src/main/java/com/example/jpa/reply/domain/Reply.java:
```diff
@@ -0,0 +1,48 @@
+    @Column(name = "reply_id")
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "board_id")
+    private Board board;
```
양방향 매핑 지양

stophwan replied:
왜 양방향 매핑을 지양해야 할까요? 특히 Jpa엔티티와 도메인이 같은 객체일 경우에는 일어날 수 있는 문제점이 정말 많습니다.
객체간 순환참조가 될 뿐만 아니라, 조금만 도메인이 많아져도 양방향 지옥에 빠져 어려움을 겪을 확률이 너무너무 큽니다... 관리도 안될 뿐더러 의도치 않는 쿼리가 나갈 수도 있고요.
정말 두 객체가 서로 알고 있어야하는지 고민해볼 필요가 있을 것 같아요.

dojinyou commented on src/main/java/com/example/jpa/member/service/MemberService.java:
```diff
@@ -18,16 +23,46 @@ public class MemberService {
+
+    public Member checkExist(Long id) {
         Member member = memberRepository.findById(id)
                 .orElseThrow(() -> new NoSuchElementException("멤버가 존재하지 않습니다."));
-
-        return MemberResponse.from(member);
+        return member;
     }
```
public이 적절한지 고민해보기.
check 메서드의 동작 방식 통일하기(조건 실패시 에러 반환, boolean 반환 등)

gyuchangShim replied:
모든 checkExist 메서드의 동작을 객체 반환으로 통일했습니다. 해당 메서드는 MemberService 뿐만 아니라 Board, Reply의 동작에도 사용해서 public으로 만들었습니다!!

stophwan reviewed (approved):
👍 고생하셨습니다!! 코멘트들을 잘 확인해서 다음 미션때 반영해봐요 ㅎㅎ

gyuchangShim committed:
Refactor: Builder 구조 변경
Board 엔티티의 of 메서드 대신 BoardCreateRequest에 toEntity 메서드로 대체 & MemberService의 checkDuplicate 메서드를 if -> Optional로 변경

gyuchangShim committed:
Refactor: Builder 구조 변경 & DTO 검증 조건 추가
Reply 엔티티 생성 과정에서 of 메서드 대신 DTO에 toEntity 메서드 추가 & Create, Update시 사용되는 DTO 객체의 속성들 검증 조건 추가
